src/main/java/com/google/devtools/build/lib/analysis/BuildView.java

```diff
@@ -428,12 +428,21 @@ public AnalysisResult update(
       logger.atInfo().log(msg);
     }
 
-    Set<ConfiguredTarget> targetsToSkip =
-        new TopLevelConstraintSemantics(
-                skyframeExecutor.getPackageManager(),
-                input -> skyframeExecutor.getConfiguration(eventHandler, input),
-                eventHandler)
-            .checkTargetEnvironmentRestrictions(skyframeAnalysisResult.getConfiguredTargets());
+    ImmutableSet.Builder<ConfiguredTarget> badTargets = ImmutableSet.builder();
+    ImmutableSet.Builder<ConfiguredTarget> badButRequestedTargets = ImmutableSet.builder();
+
+    for (ConfiguredTarget target : skyframeAnalysisResult.getConfiguredTargets()) {
+      if (target.getProvider(IncompatiblePlatformProvider.class) != null) {
+        if (requestedTargetPatterns.contains(target.getLabel().toString())) {
+          badButRequestedTargets.add(target);
```

Comment:
Can we eliminate the need for the new `targetsToFail` concept with this approach?

1. Before this change, the code `skyframeBuildView.configureTargets` (a few lines above) returns a `SkyframeAnalysisResult`
1. That class includes the relevant state `hasAnalysisError` and `configuredTargets`. Note that only includes non-error targets (I verified this with a debugger).
1. That gets instantiated at https://github.com/bazelbuild/bazel/blob/b442fca8156c68eeb93ebce454dee9c5faf44044/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeBuildView.java#L541-L545
1. If `--keep_going=false`, any errors trigger a `ViewCreationFailedException`: https://github.com/bazelbuild/bazel/blob/b442fca8156c68eeb93ebce454dee9c5faf44044/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeBuildView.java#L495 which immediately fails the build.
1. Else if `--keep_going=true`, Bazel prints a warning https://github.com/bazelbuild/bazel/blob/b442fca8156c68eeb93ebce454dee9c5faf44044/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeBuildView.java#L737-L741 and the offending target is never included in `targetsToBuild` as noted in point 2.

Your logic doesn't quite fit in because the above assumes an "error" happens when the target actively fails. But for your current approach the target just publishes an `IncompatiblePlatformProvider` and still hasn't decided if that should be an error or a skip.

So can you replicate the above logic right here? `SkyframeBuildView.processErrors`, which handles a lot of the above logic, is complex. But most of the complexity is around finding cycles and root causes to the error (like when the target itself is fine but one of its deps had the problem), none of which you need to concern yourself with here.

So can you expand your logic here to:

1. if `--keep_going=false`, for the first target in `requestedTargetPatterns`: throw a `ViewCreationFailedException` similar to 4 above. This immediately fails the build
1. else if `--keep_going=true`, print a warning similar to 5 above, then re-instantiate `skyframeAnalysisResult` with `hasAnalysisError=true` and `configuredTargets` stripped of whatever's in `badButRequestedTargets`.
1. To make 2. work you'd probably have to shift this code up before https://github.com/bazelbuild/bazel/blob/7a3eb4f61c65e9ba4a8f9be68867f40dc5166845/src/main/java/com/google/devtools/build/lib/analysis/BuildView.java#L418, which is already starting to read the results.

Reply:
Trying to summarize the above succinctly:

1. Bazel already has the concept of `targetsToBuild` and `targetsToSkip`, and implicitly considers anything *not* in either of those as `targetsToFail`.
1. It does that by making sure anything that doesn't go into `targetsToBuild` either a) breaks the build immediately with an error (when `--keep_going=false`) or b) prints a warning
1. You should be able to replicate that exact logic here as a simple post-processing step: using the extra information you care about (is the target explicit? does it have an `IncompatiblePlatformProvider`?), further filter results out of `targetsToBuild` and make sure you message the error accordingly.

Reply:
That's beautiful; it did not occur to me to think of the targets being bucketed that way. Thank you!

I had to add a helper function to `SkyframeAnalysisResult` to filter out incompatible targets since its constructor is not public. However, the concept of "targetsToFail" is now gone. I definitely like this version quite a bit better.

Still working on figuring out the other issue bazel is complaining about in regards to the "These artifacts do not have a generating action" error.